src/core.rs

```diff
@@ -671,30 +906,60 @@ const HYPHEN_PENALTY: i32 = 25;
 /// by Knuth and Plass. The implementation here is based on [Python
 /// code by David
 /// Eppstein](https://github.com/jfinkels/PADS/blob/master/pads/wrap.py).
-pub fn wrap_optimal_fit<'a, T: Fragment, F: Fn(usize) -> usize>(
-    fragments: &'a [T],
-    line_widths: F,
-) -> Vec<&'a [T]> {
-    let mut widths = Vec::with_capacity(fragments.len() + 1);
-    let mut width = 0;
-    widths.push(width);
-    for fragment in fragments {
-        width += fragment.width() + fragment.whitespace_width();
-        widths.push(width);
+pub fn wrap_optimal_fit<F, W>(
+    fragments: F,
+    line_widths: W,
+) -> WrapOptimalFit<<F as IntoIterator>::Item>
+where
+    F: IntoIterator,
+    <F as IntoIterator>::Item: Fragment,
+    W: Fn(usize) -> usize,
+{
+    struct FragmentData<F> {
+        /// The fragment itself.
+        fragment: F,
+        /// The cumulative width of all the fragments up to and including this one.
+        cu_width: usize,
+        /// The whitespace width of this fragment.
+        whitespace_width: usize,
+        /// The penalty width of this fragment.
+        penalty_width: usize,
     }
 
+    // We precompute some fragment data for use later on.
+    let mut cu_width = 0;
+    let fragments: Vec<_> = fragments
+        .into_iter()
+        .map(|fragment| {
+            let whitespace_width = fragment.whitespace_width();
+            let penalty_width = fragment.penalty_width();
+            cu_width += fragment.width() + whitespace_width;
+            FragmentData {
+                fragment,
+                cu_width,
+                whitespace_width,
+                penalty_width,
+            }
+        })
+        .collect();
+
     let line_numbers = LineNumbers::new(fragments.len());
 
-    let minima = smawk::online_column_minima(0, widths.len(), |minima, i, j| {
+    let minima = smawk::online_column_minima(0, fragments.len() + 1, |minima, i, j| {
         // Line number for fragment `i`.
         let line_number = line_numbers.get(i, &minima);
         let target_width = std::cmp::max(1, line_widths(line_number));
 
+        // The last fragment of the range fragments[i..j].
+        let last_fragment = &fragments[j - 1];
+
         // Compute the width of a line spanning fragments[i..j] in
         // constant time. We need to adjust widths[j] by subtracting
         // the whitespace of fragment[j-i] and then add the penalty.
-        let line_width = widths[j] - widths[i] - fragments[j - 1].whitespace_width()
-            + fragments[j - 1].penalty_width();
+        let line_width = last_fragment.cu_width
+            - i.checked_sub(1).map(|i| fragments[i].cu_width).unwrap_or(0)
```

Comment:
I think this deserves a comment :-) Or rather, please pull it out into a local variable.

I guess you're basically differentiating between `i == 0` (which is mapped to `0`) and `i > 0` which is mapped to `fragments[i].cu_width`? Why is this needed? Before I computed `widths` so that I could always index into it with `i` and `j`, but I guess that changed?